### PR TITLE
Improve drive detection and UI sync to deal with duplicate entries

### PR DESCRIPTION
**Pull request by AndyMDoyle**
#3231 

Primary change is making the app service connection a singleton which means multiple messages can be sent without issues launching additional processes. A lot of this has been taken from a @gave92's network drives PR and should make that PR a smaller update.

The logic to add local drives and cloud drives to the side bar has also been refactored to ensure this is done on the UI thread and that a lock is obtained to prevent multiple updates to the side bar items at the same time. Now a lot of the code is async, this is done using a `SemaphoreSlim` to ensure one thread at a time updates the side bar items.

So far this update has been 100% reliable in debug mode, release mode, and with SharePoint sync folder support on top which results in an additional concurrent full trust call at startup.

### CODE REVIEW

File: Files/Views/MainPage.xaml.cs [hunk 1]
```diff
@@ -51,6 +53,7 @@ public TabItem SelectedTabItem
 
         public static ObservableCollection<TabItem> AppInstances = new ObservableCollection<TabItem>();
         public static ObservableCollection<INavigationControlItem> SideBarItems = new ObservableCollection<INavigationControlItem>();
+        public static SemaphoreSlim sideBarItemsSemaphore = new SemaphoreSlim(1, 1);
```
Comment (gave92) [inline Files/Views/MainPage.xaml.cs:56]: `SideBarItemsSemaphore ` since it's public

File: Files/Filesystem/Drives.cs [hunk 2]
```diff
@@ -118,29 +110,39 @@ private async Task RefreshUI()
                 // Defer because UI-thread is not ready yet (and DriveItem requires it?)
                 CoreApplication.MainView.Activated += RefreshUI;
             }
+            Logger.Info("RefreshUI() complete");
```
Comment (gave92) [inline Files/Filesystem/Drives.cs:113]: I'd probably remove all this logs

**Review by gave92 — CHANGES_REQUESTED** [review]

File: Files/Filesystem/Drives.cs (hunk 2 above)
Comment (AndyMDoyle) [reply Files/Filesystem/Drives.cs:113]: Should be able to do without them. Easy enough to add back later for troubleshooting should they be needed for testing.

**Comment by gave92** [comment]
Still duplicates :(
image
Wrong drive order:
image

**Comment by AndyMDoyle** [comment]
@gave92, PR updated to remove drive ordering. I've realised I didn't see that because my drives are named in a way they end up in order of the drive letters.

Can you try and repro the duplication issue by swapping

```
_ = Task.Factory.StartNew(async () =>
{
    await DrivesManager.EnumerateDrivesAsync();
    await CloudDrivesManager.EnumerateDrivesAsync();
});
``` 

for 

```
await DrivesManager.EnumerateDrivesAsync();
await CloudDrivesManager.EnumerateDrivesAsync();
```

Task.Factory.StartNew doesn't preserve the synchronization context and has caused me issues in the past. This way the drive lists should be fully populated before MainPage is constructed.

**Comment by gave92** [comment]
@AndyMDoyle yeah but I really don't want to wait for the sidebar to be fully populated before showing MainPage :)
I'm pushing here what fixes the duplicate issue *for me*.

**Comment by AndyMDoyle** [comment]
Neither do I really but doing that did get rid of duplication issue for me.

Your commit looks great and I can't see any issues on my side. One question though, the existing usage of `BulkConcurrentObservableCollection` in `ItemViewModel` calls the `BeginBulkOperation()` method before working on the collection, followed by `EndBulkOperation()`. Should `BeginBulkOperation()` be called from these new places?

And in `ItemViewModel()`, `EndBulkOperation()` is called on the UI thread, so is that also needed instead of that happening on a background thread?

**Comment by gave92** [comment]
Yes, EndBulkOperation() must be called on the UI thread, I should be already doing it, if not please correct :)
BeginBulkOperation() is not really mandatory in this case. It's used to pause firing events for collection changes but here I just want to call an additional `NotifyCollectionChangedAction.Reset` event when I'm done adding items to workaround the duplicate issue.
> Your commit looks great and I can't see any issues on my side
I can't believe that xD If you minimize and restore the app *withou being attached to the debugger* are you really not getting a crash?

**Comment by AndyMDoyle** [comment]
@gave92, yes I am seeing a crash related to BulkConcurrentObservableCollection.OnCollectionChanged if I run without the debugger.

Trying the latest commit now.

**Comment by gave92** [comment]
Hi @hez2010. In this PR we were trying to switch the MainPage SidebarItems list from ObservableCollection to the BulkConcurrentObservableCollection. Unfortunately we were getting a crash anytime the `Remove` method was called with the following error: `Parameter name: index|System.ArgumentOutOfRangeException: This collection cannot work with indices larger than Int32.MaxValue - 1 (0x7FFFFFFF - 1).`.

Do you understand why?

If you wish to test you can checkout commit 5eceb15df6c7a78e25bb654d157637cb2d5c02cf (not the latest one). To reproduce the bug you can suspend/resume the app (or minimize/restore it with no debugger attached).

**Comment by AndyMDoyle** [comment]
@gave92 everything looks okay when using an ObservableCollection. Release mode starts up nice an quick and the local drives and cloud drives sections appear when ready.

**Comment by hez2010** [comment]
@AndyMDoyle I've proposed a fix to `BulkConcurrentObservableCollection`, could I push the fix to your branch?

**Comment by hez2010** [comment]
I think this PR is ready for merge. @gave92 could you test again to see if there're still issues remaining or performance degradation about `BulkConcurrentObservableCollection`?

**Comment by hez2010** [comment]
I think this PR is ready for merge. @gave92 could you test again to see if there're still issues remaining or performance degradation about `BulkConcurrentObservableCollection`?

**Review by gave92 — APPROVED** [review]

**Comment by yaichenbaum** [comment]
@AndyMDoyle Great work!

**Pull request merged**